ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/binary_ng_device_operation.cpp

```diff
@@ -0,0 +1,232 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "binary_ng_device_operation.hpp"
+
+namespace ttnn::operations::binary_ng {
+
+SubtileBroadcastType get_subtile_broadcast_type(uint32_t a_h, uint32_t a_w, uint32_t b_h, uint32_t b_w) {
+    if (a_h == b_h && a_w == b_w) {
+        return SubtileBroadcastType::NONE;
+    }
+    if (a_h == 1 && a_w == 1) {
+        return SubtileBroadcastType::SCALAR_A;
+    }
+    if (b_h == 1 && b_w == 1) {
+        return SubtileBroadcastType::SCALAR_B;
+    }
+    if (a_h == 1 /* && a_w != 1 */ && b_w == 1 /* && b_h != 1 */) {
+        return SubtileBroadcastType::ROW_A_COL_B;
+    }
+    if (a_w == 1 /* && a_h != 1 */ && b_h == 1 /* && b_w != 1 */) {
+        return SubtileBroadcastType::ROW_B_COL_A;
+    }
+    if (a_h == 1) {
+        return SubtileBroadcastType::ROW_A;
+    }
+    if (a_w == 1) {
+        return SubtileBroadcastType::COL_A;
+    }
+    if (b_h == 1) {
+        return SubtileBroadcastType::ROW_B;
+    }
+    if (b_w == 1) {
+        return SubtileBroadcastType::COL_B;
+    }
+
+    TT_THROW("Invalid subtile broadcast type");
+}
+
+tt::stl::hash::hash_t BinaryNgDeviceOperation::operation_attributes_t::to_hash() const {
+    return tt::stl::hash::hash_objects_with_default_seed(
+        binary_op_type, memory_config, dtype, compute_kernel_config, subtile_broadcast_type);
+}
+
+void BinaryNgDeviceOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    // We don't support sharding for now
+    const auto& input_tensor_a = tensor_args.input_tensor_a;
+    const auto& input_tensor_b = tensor_args.input_tensor_b;
+    const auto& output_tensor = tensor_args.output_tensor;
+
+    TT_FATAL(
+        input_tensor_b.has_value() != attributes.scalar.has_value(), "Either the tensor b or scalar should be set");
+
+    BinaryNgDeviceOperation::validate_on_program_cache_hit(attributes, tensor_args);
+
+    TT_FATAL(input_tensor_a.get_layout() == Layout::TILE, "Input to eltwise binary must be tilized");
+    TT_FATAL(
+        input_tensor_a.memory_config().memory_layout == TensorMemoryLayout::INTERLEAVED,
+        "Only interleaved layout is supported");
+    TT_FATAL(
+        attributes.memory_config.memory_layout == TensorMemoryLayout::INTERLEAVED,
+        "Only interleaved layout is supported");
+
+    if (input_tensor_b.has_value()) {
+        TT_FATAL(input_tensor_b->get_layout() == Layout::TILE, "Input to eltwise binary must be tilized");
```

Comment:
would be good to make it clear that in this case the issue is with tensor B, and in the first case with tensor A